src/durable-cache/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! The crate provides a durable key-value cache abstraction implemented by persist.
+
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+use std::sync::Arc;
+
+use differential_dataflow::consolidation::consolidate_updates;
+use mz_ore::collections::AssociativeExt;
+use mz_persist_client::error::UpperMismatch;
+use mz_persist_client::read::{ListenEvent, Subscribe};
+use mz_persist_client::write::WriteHandle;
+use mz_persist_client::{Diagnostics, PersistClient};
+use mz_persist_types::{Codec, ShardId};
+use timely::progress::Antichain;
+
+pub trait DurableCacheCodec {
```

Comment:
I'm curious how things shake out when you hook this all up. Mentioning again that I'm happy for us to rip out `DurableCacheCodec` and hardcode `<SourceData, (), mz_repr::Timestamp, i64>` plus the one json column RelationDesc schema

Reply:
If you're curious I have a WIP here: https://github.com/jkosh44/materialize/commit/7a9ed0fc9d29bde05447ae1d07c652e5317128e3. I was actually interested in your take on using `<SourceData, SourceData, mz_repr::Timestamp, i64>` instead. It ended up making the implementation easier because I didn't need to implement `Ord` on the value, but I had to remove the call to `consolidate` in `set_many`.

In terms of `DurableCacheCodec`, I sort of like it because it allows more flexibility and the boilerplate doesn't feel that oppressive, but also I don't feel strongly either way so I'm happy to remove it.